### PR TITLE
Add stft op

### PR DESCRIPTION
**Pull request by small1945**
一. 涉及到` fft `运算部分
- cpu kernel 底层` fft `运算采用`pocketfft_hdronly.h` 库（外部引入库文件）
- cuda kernel 底层 `fft `运算采用 cuda 自带的`cufft`库 （只支持cuda 11.0以上版本）
- 针对fft运算封装好对应的类，后续根据需要可进行扩展

二. `stft `算子相关工作
- 未支持参数：return_complex

### CODE REVIEW

**Comment by doombeaker** [comment]
我想用这个分支做测试。但是编译时报了一些由 clang-tidy 检查出的错误导致编译不通过，大多是数据类型隐式转换的 warning。

智敏可以帮忙看看修改下吗。或者说这个问题是已知的，可以不修改代码的前提下，用其它方法绕过，编译成功？如果有，也请告诉我，谢谢。

```text
/home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp: In member function ‘oneflow::Maybe<oneflow::one::Tensor> oneflow::one::functional::impl::StftFunctor::operator()(const std::shared_ptr<oneflow::one::Tensor>&, int64_t, oneflow::Optional<long int>, oneflow::Optional<long int>, const oneflow::Optional<oneflow::one::Tensor>&, bool, const string&, bool, bool, bool) const’:
/home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:62: error: narrowing conversion of ‘(int64_t)n_fft’ from ‘int64_t’ {aka ‘long int’} to ‘int’ [-Werror=narrowing]
 3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
      |                                                              ^~~~~
/home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
  105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
      |                                        ^~~~~~~~~~~
/home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
 3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
      |         ^~~~
/home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:62: error: narrowing conversion of ‘n_fft’ from ‘const int64_t’ {aka ‘const long int’} to ‘int’ [-Werror=narrowing]
 3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
      |                                                              ^~~~~
/home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
  105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
      |                                        ^~~~~~~~~~~
/home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
 3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
      |         ^~~~
/home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3182:61: error: narrowing conversion of ‘(new_hop_length * ((int64_t)strides.std::vector<int>::at(1)))’ from ‘int64_t’ {aka ‘long int’} to ‘int’ [-Werror=narrowing]
 3182 |                              {strides.at(0), new_hop_length * strides.at(1), strides.at(1)}, 0));
      |                                              ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
/home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
  105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
      |                                        ^~~~~~~~~~~
/home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
 3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
      |         ^~~~
/home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3182:61: error: narrowing conversion of ‘(new_hop_length * ((int64_t)strides.std::vector<int>::at(1)))’ from ‘int64_t’ {aka ‘long int’} to ‘int’ [-Werror=narrowing]
 3182 |                              {strides.at(0), new_hop_length * strides.at(1), strides.at(1)}, 0));
      |                                              ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
/home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
  105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
      |                                        ^~~~~~~~~~~
/home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
 3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
```

**Comment by small1945** [comment]
> 我想用这个分支做测试。但是编译时报了一些由 clang-tidy 检查出的错误导致编译不通过，大多是数据类型隐式转换的 warning。
> 
> 智敏可以帮忙看看修改下吗。或者说这个问题是已知的，可以不修改代码的前提下，用其它方法绕过，编译成功？如果有，也请告诉我，谢谢。
> 
> ```
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp: In member function ‘oneflow::Maybe<oneflow::one::Tensor> oneflow::one::functional::impl::StftFunctor::operator()(const std::shared_ptr<oneflow::one::Tensor>&, int64_t, oneflow::Optional<long int>, oneflow::Optional<long int>, const oneflow::Optional<oneflow::one::Tensor>&, bool, const string&, bool, bool, bool) const’:
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:62: error: narrowing conversion of ‘(int64_t)n_fft’ from ‘int64_t’ {aka ‘long int’} to ‘int’ [-Werror=narrowing]
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |                                                              ^~~~~
> /home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
>   105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
>       |                                        ^~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |         ^~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:62: error: narrowing conversion of ‘n_fft’ from ‘const int64_t’ {aka ‘const long int’} to ‘int’ [-Werror=narrowing]
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |                                                              ^~~~~
> /home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
>   105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
>       |                                        ^~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |         ^~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3182:61: error: narrowing conversion of ‘(new_hop_length * ((int64_t)strides.std::vector<int>::at(1)))’ from ‘int64_t’ {aka ‘long int’} to ‘int’ [-Werror=narrowing]
>  3182 |                              {strides.at(0), new_hop_length * strides.at(1), strides.at(1)}, 0));
>       |                                              ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
>   105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
>       |                                        ^~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |         ^~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3182:61: error: narrowing conversion of ‘(new_hop_length * ((int64_t)strides.std::vector<int>::at(1)))’ from ‘int64_t’ {aka ‘long int’} to ‘int’ [-Werror=narrowing]
>  3182 |                              {strides.at(0), new_hop_length * strides.at(1), strides.at(1)}, 0));
>       |                                              ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
>   105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
>       |                                        ^~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
> ```

**Comment by small1945** [comment]
> 我想用这个分支做测试。但是编译时报了一些由 clang-tidy 检查出的错误导致编译不通过，大多是数据类型隐式转换的 warning。
> 
> 智敏可以帮忙看看修改下吗。或者说这个问题是已知的，可以不修改代码的前提下，用其它方法绕过，编译成功？如果有，也请告诉我，谢谢。
> 
> ```
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp: In member function ‘oneflow::Maybe<oneflow::one::Tensor> oneflow::one::functional::impl::StftFunctor::operator()(const std::shared_ptr<oneflow::one::Tensor>&, int64_t, oneflow::Optional<long int>, oneflow::Optional<long int>, const oneflow::Optional<oneflow::one::Tensor>&, bool, const string&, bool, bool, bool) const’:
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:62: error: narrowing conversion of ‘(int64_t)n_fft’ from ‘int64_t’ {aka ‘long int’} to ‘int’ [-Werror=narrowing]
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |                                                              ^~~~~
> /home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
>   105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
>       |                                        ^~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |         ^~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:62: error: narrowing conversion of ‘n_fft’ from ‘const int64_t’ {aka ‘const long int’} to ‘int’ [-Werror=narrowing]
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |                                                              ^~~~~
> /home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
>   105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
>       |                                        ^~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |         ^~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3182:61: error: narrowing conversion of ‘(new_hop_length * ((int64_t)strides.std::vector<int>::at(1)))’ from ‘int64_t’ {aka ‘long int’} to ‘int’ [-Werror=narrowing]
>  3182 |                              {strides.at(0), new_hop_length * strides.at(1), strides.at(1)}, 0));
>       |                                              ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
>   105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
>       |                                        ^~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
>       |         ^~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3182:61: error: narrowing conversion of ‘(new_hop_length * ((int64_t)strides.std::vector<int>::at(1)))’ from ‘int64_t’ {aka ‘long int’} to ‘int’ [-Werror=narrowing]
>  3182 |                              {strides.at(0), new_hop_length * strides.at(1), strides.at(1)}, 0));
>       |                                              ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/common/just.h:105:40: note: in definition of macro ‘__JustStackCheckWrapper__’
>   105 | #define __JustStackCheckWrapper__(...) __VA_ARGS__
>       |                                        ^~~~~~~~~~~
> /home/yaochi/oneflow/oneflow/core/functional/impl/math_functor.cpp:3181:9: note: in expansion of macro ‘JUST’
>  3181 |         JUST(view::AsStrided(input_tensor, {batch, n_frames, n_fft},
> ```

已修改代码，可以拉取最新的分支试试
我这边本地用的是 gcc 编译，一些警告并没有当错误处理，所以编译能通过

File: oneflow/core/functional/impl/math_functor.cpp [hunk 1]
```diff
@@ -3224,6 +3224,113 @@ class InplaceAddCDivFunctor {
...
+    if (new_win_length < n_fft) {
+      temp_tensor = JUST(functional::Fill(temp_tensor, 0));
+      auto left = (n_fft - new_win_length) / 2;
+
+      if (window.has_value()) {
+        // TODO(yzm):Copy the window matrix to the defined range,such as
+        //'''
+        //      functional::AssignLocalTensor(JUST(functional::Narrow(temp_tensor, 0,
+        //      left,new_win_length)), window);
+        //'''
```
Comment (wyg1997) [inline oneflow/core/functional/impl/math_functor.cpp:3313]: 这里应该给报错吧，不能直接忽略

File: oneflow/core/functional/impl/math_functor.cpp [hunk 2]
```diff
@@ -3224,6 +3224,113 @@ class InplaceAddCDivFunctor {
...
+      temp_tensor = JUST(window);
+      CHECK_OR_RETURN(temp_tensor->shape()->NumAxes() == 1
+                      && temp_tensor->shape()->at(0) == new_win_length)
+          << Error::RuntimeError()
+          << "Expected a 1D window tensor of size equal to win_length=" << new_win_length
+          << ", but got window with size " << temp_tensor->shape()->ToString();
+    }
+    if (new_win_length < n_fft) {
+      temp_tensor = JUST(functional::Fill(temp_tensor, 0));
+      auto left = (n_fft - new_win_length) / 2;
```
Comment (wyg1997) [inline oneflow/core/functional/impl/math_functor.cpp:3306]: 这里简单的数据类型就不要用 auto 了

File: oneflow/ir/include/OneFlow/OneFlowUserOps.td [hunk 3]
```diff
@@ -4924,6 +4924,31 @@ def OneFlow_ErfInvOp : OneFlow_BaseOp<"erfinv", [NoSideEffect, DeclareOpInterfac
...
+    DefaultValuedAttr<StrAttr, "\"reflect\"">:$pad_mode,
+    DefaultValuedAttr<BoolAttr, "false">:$normalized,
+    DefaultValuedAttr<BoolAttr, "false">:$onesided,
+    DefaultValuedAttr<BoolAttr, "false">:$return_complex
+  );
+
+  let has_logical_tensor_desc_infer_fn = 1;
+  let has_physical_tensor_desc_infer_fn = 1;
+  let has_get_sbp_fn = 1;
+  let has_data_type_infer_fn = 1;
```
Comment (wyg1997) [inline oneflow/ir/include/OneFlow/OneFlowUserOps.td:4950]: 可以支持一个 GetComputeComplexity 函数，大概估计出这个计算的复杂度

File: oneflow/ir/include/OneFlow/OneFlowUserOps.td [hunk 4]
```diff
@@ -11439,3 +11464,5 @@ def OneFlow_MlirJitOp : OneFlow_JITLikeOp<"mlir_jit"> {}
 def OneFlow_KernelLaunchOp : OneFlow_JITLikeOp<"kernel_launch"> {}
 
 #endif // GET_ONEFLOW_MLIR_JIT_OP_DEFINITIONS
+
+
```
Comment (wyg1997) [inline oneflow/ir/include/OneFlow/OneFlowUserOps.td:11468]: 这两行可以直接删了

File: oneflow/user/kernels/cufftplancache.h [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "oneflow/core/framework/framework.h"
+#include "oneflow/core/kernel/new_kernel_util.h"
+#include "oneflow/core/ep/cuda/cuda_stream.h"
+#include <cufft.h>
+#include <cufftXt.h>
```
Comment (wyg1997) [inline oneflow/user/kernels/cufftplancache.h:21]:
```suggestion
#include <cufft.h>
#include <cufftXt.h>
#include "oneflow/core/framework/framework.h"
#include "oneflow/core/kernel/new_kernel_util.h"
#include "oneflow/core/ep/cuda/cuda_stream.h"
```

File: oneflow/user/kernels/cufftplancache.h [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+struct CuFFtParams {
+  int32_t ndim;
+  int32_t output_shape[max_rank + 1];
+  int32_t input_shape[max_rank + 1];
+  int32_t input_strides[max_rank + 1];
+  int32_t output_strides[max_rank + 1];
+  int32_t* rank;
+  int32_t batch;
+  CuFFtParams() = default;
```
Comment (wyg1997) [inline oneflow/user/kernels/cufftplancache.h:37]: 这个默认构造函数有用到吗？没有的话可以删掉

File: python/oneflow/test/modules/test_stft.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            normalized=normalized_value,
+        )
+        return y
+
+    def test_stft_with_2D_random_data(test_case):
+        min_cuda_version = int(re.search("\d{2}", flow.__version__).group())
+        if min_cuda_version < 11:
+            device = cpu_device()
+        else:
+            device = random_device()
```
Comment (wyg1997) [inline python/oneflow/test/modules/test_stft.py:81]: 这部分可以加个注释，说明一下 cuda 低版本不支持 cufft

File: oneflow/core/functional/impl/math_functor.cpp (hunk 1 above)
Comment (small1945) [reply oneflow/core/functional/impl/math_functor.cpp:3313]: done

File: oneflow/core/functional/impl/math_functor.cpp (hunk 2 above)
Comment (small1945) [reply oneflow/core/functional/impl/math_functor.cpp]: done

File: oneflow/ir/include/OneFlow/OneFlowUserOps.td (hunk 3 above)
Comment (small1945) [reply oneflow/ir/include/OneFlow/OneFlowUserOps.td:4950]: done,已添加 todo

File: oneflow/ir/include/OneFlow/OneFlowUserOps.td (hunk 4 above)
Comment (small1945) [reply oneflow/ir/include/OneFlow/OneFlowUserOps.td]: done

File: python/oneflow/test/modules/test_stft.py (hunk 7 above)
Comment (small1945) [reply python/oneflow/test/modules/test_stft.py:81]: done

File: oneflow/user/kernels/cufftplancache.h (hunk 6 above)
Comment (small1945) [reply oneflow/user/kernels/cufftplancache.h]: done

File: oneflow/core/functional/impl/math_functor.cpp [hunk 8]
```diff
@@ -3224,6 +3224,118 @@ class InplaceAddCDivFunctor {
...
+      if (window.has_value()) {
+        // TODO(yzm):Copy the window matrix to the defined range,such as
+        //'''
+        //      functional::AssignLocalTensor(JUST(functional::Narrow(temp_tensor, 0,
+        //      left,new_win_length)), window);
+        //'''
+        // Remove the following check after support
+        CHECK_OR_RETURN(false) << Error::RuntimeError()
+                               << "The following conditions are not currently supported: "
+                                  "win_length<n_fft and the window function is customized";
```
Comment (wyg1997) [inline oneflow/core/functional/impl/math_functor.cpp:3317]: 可以直接 `return Error::RuntimeError` 的

File: oneflow/user/ops/stft_op.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+/* static */ Maybe<void> StftOp::InferDataType(user_op::InferContext* ctx) {
+  ctx->SetOutputDType("output", 0, ctx->InputDType("input", 0));
+  return Maybe<void>::Ok();
+}
+
+/*static*/ Maybe<double> StftOp::GetComputeComplexity(user_op::ComputeComplexityFnContext* ctx) {
+  // TODO: add ComputeComplexityFun
+  return 0.0;
```
Comment (wyg1997) [inline oneflow/user/ops/stft_op.cpp:79]: 这里可以大概估计一下，计算量不是线性的，但也不能轻易设置成 0

File: python/oneflow/test/modules/test_stft.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    pow = np.random.randint(2, 10)
+    result = 1
+    for i in range(pow):
+        result = result * 2
+    return result
+
+
+class TestStft(flow.unittest.TestCase):
+    @autotest(
+        n=20, check_graph=False, check_grad_use_random_data=False, auto_backward=False,
```
Comment (wyg1997) [inline python/oneflow/test/modules/test_stft.py:48]: 这个测试要跑多久，如果比较长的话，这个可以只循环 5 次，大概控制在 10-30s

**Review by wyg1997 — APPROVED** [review]

**Review by BBuf — APPROVED** [review]

**Review by Ldpe2G — APPROVED** [review]

**Pull request merged**